src/core/providers/ogr/qgsogrlayermetadataprovider.cpp

```diff
@@ -0,0 +1,63 @@
+/***************************************************************************
+  qgsogrlayermetadataprovider.cpp - QgsOgrLayerMetadataProvider
+
+ ---------------------
+ begin                : 24.8.2022
+ copyright            : (C) 2022 by Alessandro Pasotti
+ email                : elpaso at itopen dot it
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#include "qgsogrlayermetadataprovider.h"
+#include "qgsprovidermetadata.h"
+#include "qgsproviderregistry.h"
+#include "qgsfeedback.h"
+#include "qgsabstractdatabaseproviderconnection.h"
+
+
+QString QgsOgrLayerMetadataProvider::id() const
+{
+  return QStringLiteral( "ogr" );
+}
+
+QgsLayerMetadataSearchResult QgsOgrLayerMetadataProvider::search( const QgsMetadataSearchContext &searchContext, const QString &searchString, const QgsRectangle &geographicExtent, QgsFeedback *feedback ) const
+{
+  QgsLayerMetadataSearchResult results;
+  QgsProviderMetadata *md { QgsProviderRegistry::instance()->providerMetadata( id( ) ) };
+
+  if ( md && ( ! feedback || ! feedback->isCanceled( ) ) )
+  {
+    const QMap<QString, QgsAbstractProviderConnection *> cConnections { md->connections( ) };
+    for ( const QgsAbstractProviderConnection *conn : std::as_const( cConnections ) )
+    {
+
+      if ( feedback && feedback->isCanceled() )
+      {
+        break;
+      }
+
+      if ( const QgsAbstractDatabaseProviderConnection *dbConn = static_cast<const QgsAbstractDatabaseProviderConnection *>( conn ) )
+      {
+        try
+        {
+          const QList<QgsLayerMetadataProviderResult> res { dbConn->searchLayerMetadata( searchContext, searchString, geographicExtent, feedback ) };
+          for ( const QgsLayerMetadataProviderResult &result : std::as_const( res ) )
+          {
+            results.addMetadata( result );
+          }
+        }
+        catch ( const QgsProviderConnectionException &ex )
+        {
+          results.addError( QObject::tr( "An error occurred while searching for metadata in connection %1: %2" ).arg( conn->uri(), ex.what() ) );
```

Comment:
Just thinking -- wouldn't it be better to have a capability flag we can check on the database connection prior to trying to obtain results? That would allow us to differentiate errors which occurred because a connection which SHOULD be able to search metadata failed vs a connection which can't be searched for metadata and nothing went wrong.

Reply:
It's complicated: the pre-existing implementation of md storage was for GPKG only (delegated to GDAL) and it was working transparently with implicit support for md storage. For PG (and potentially other real DBs) I introduced an option to flag the connection as suitable for md storage (same as project/style storage).

So, for PG we actually have a flag in the connection that explicitly tells us if the connection is supposed to support md storage, for GPKG we don't, IMHO is doesn't make much sense to add it to GPKG because it is local file-based.

I'll see if I can make error reporting much precise though.

Reply:
I checked and I think we are good: in PG the error is raised when the connection is supposed to support md storage but something goes wrong, GPKG is supposed to always support md storage so we raise the error all the times.